File: KargoPaylasimSwiftUI/KargoPaylasimSwiftUI/viewmodel/HomeVM.swift
//
//  HomeVMv.swift
//  KargoPaylasimSwiftUI
//
//  Created by Alperen Saraç on 30.01.2026.
//

import Foundation

@MainActor
final class HomeVM: ObservableObject {
    @Published var shipments: [Shipment] = []
    @Published var addresses: [Address] = []
    @Published var isLoading = false
    @Published var errorText: String?

    private let api: APIClient
    init(api: APIClient) { self.api = api }

    func refresh() async {
        isLoading = true
        defer { isLoading = false }
        errorText = nil

        do {
            async let s: ApiResp<ShipmentListData> = api.get(
                .shipmentList(type: "all"),
                query: ["type": "all"],
                as: ApiResp<ShipmentListData>.self
            )

            async let a: ApiResp<AddressListData> = api.get(
                .addressList,
                as: ApiResp<AddressListData>.self
            )

            let ship = try await s
            let addr = try await a

            guard ship.ok else { throw APIError.server(ship.error ?? "shipment_list failed") }
            guard addr.ok else { throw APIError.server(addr.error ?? "address_list failed") }

            shipments = ship.data?.items ?? []
            addresses = addr.data?.items ?? []
        } catch {
            errorText = error.localizedDescription
        }
    }

    func setDefaultAddress(_ id: Int) async {
        isLoading = true
        defer { isLoading = false }
        errorText = nil

        do {
            struct Req: Encodable { let id: Int }

            let res: ApiResp<Bool> = try await api.postJSON(
                .addressSetDefault,
                body: Req(id: id),
                as: ApiResp<Bool>.self
            )

            guard res.ok else { throw APIError.server(res.error ?? "Varsayılan ayarlanamadı") }

            // UI'yi anında güncelle
            addresses = addresses.map { a in
                Address(
                    id: a.id,
                    title: a.title,
                    city: a.city,
                    district: a.district,
                    address_line: a.address_line,
                    is_default: (a.id == id ? 1 : 0)
                )
            }
        } catch {
            errorText = error.localizedDescription
        }
    }

    func deleteAddress(_ id: Int) async {
        isLoading = true
        defer { isLoading = false }
        errorText = nil

        do {
            struct Req: Encodable { let id: Int }

            let res: ApiResp<Bool> = try await api.postJSON(
                .addressDelete,
                body: Req(id: id),
                as: ApiResp<Bool>.self
            )

            guard res.ok else { throw APIError.server(res.error ?? "Adres silinemedi") }

            // UI'den kaldır
            addresses.removeAll { $0.id == id }
        } catch {
            errorText = error.localizedDescription
        }
    }
}

File: flutter_kargo_paylasim/lib/viewmodel/home_vm.dart
import 'package:flutter/foundation.dart';
import '../service/api_client.dart';
import '../service/api_resp.dart';
import '../service/endpoints.dart';
import '../model/address.dart';
import '../model/shipment.dart';

class HomeVM extends ChangeNotifier {
  final ApiClient api;
  HomeVM(this.api);

  bool isLoading = false;
  String? errorText;

  List<Shipment> shipments = [];
  List<Address> addresses = [];

  Future<void> refresh() async {
    isLoading = true;
    errorText = null;
    notifyListeners();

    try {
      final shipJson = await api.getJson(Endpoints.shipmentList, query: {"type": "all"});
      final addrJson = await api.getJson(Endpoints.addressList);

      final shipResp = ApiResp.fromJson(shipJson, (d) {
        final items = (d["items"] as List? ?? []).cast<Map<String, dynamic>>();
        return items.map((e) => Shipment.fromJson(e)).toList();
      });

      final addrResp = ApiResp.fromJson(addrJson, (d) {
        final items = (d["items"] as List? ?? []).cast<Map<String, dynamic>>();
        return items.map((e) => Address.fromJson(e)).toList();
      });

      if (!shipResp.ok) throw ApiError(shipResp.error ?? "shipment_list failed");
      if (!addrResp.ok) throw ApiError(addrResp.error ?? "address_list failed");

      shipments = shipResp.data ?? [];
      addresses = addrResp.data ?? [];
    } catch (e) {
      errorText = e.toString();
    } finally {
      isLoading = false;
      notifyListeners();
    }
  }

  Future<void> setDefaultAddress(int id) async {
    isLoading = true;
    errorText = null;
    notifyListeners();

    try {
      final j = await api.postJson(Endpoints.addressSetDefault, {"id": id});
      final r = ApiResp<bool>.fromJson(j, (d) => true);
      if (!r.ok) throw ApiError(r.error ?? "Varsayılan ayarlanamadı");

      // hızlı UI update
      addresses = addresses
          .map((a) => Address(
        id: a.id,
        title: a.title,
        city: a.city,
        district: a.district,
        addressLine: a.addressLine,
        isDefault: a.id == id ? 1 : 0,
      ))
          .toList();
    } catch (e) {
      errorText = e.toString();
    } finally {
      isLoading = false;
      notifyListeners();
    }
  }

  Future<void> deleteAddress(int id) async {
    isLoading = true;
    errorText = null;
    notifyListeners();

    try {
      final j = await api.postJson(Endpoints.addressDelete, {"id": id});
      final r = ApiResp<bool>.fromJson(j, (d) => true);
      if (!r.ok) throw ApiError(r.error ?? "Adres silinemedi");

      addresses.removeWhere((a) => a.id == id);
    } catch (e) {
      errorText = e.toString();
    } finally {
      isLoading = false;
      notifyListeners();
    }
  }
}
